anaconda_project/env_spec.py

```diff
@@ -309,15 +279,46 @@ def inherit_from_names(self):
 
     def path(self, project_dir):
         """The filesystem path to the default conda env containing our packages."""
+        def _prefix(base_path):
+            base_path = os.path.expanduser(base_path) if base_path else 'envs'
+            path = os.path.abspath(os.path.join(project_dir, base_path, self.name))
+            return path
+
+        def _found(env_prefix):
+            return os.path.isdir(os.path.join(env_prefix, 'conda-meta'))
+
         if self._path is None:
-            for base in os.environ.get('ANACONDA_PROJECT_ENVS_PATH', '').split(os.pathsep):
-                base = os.path.expanduser(base) if base else 'envs'
-                path = os.path.abspath(os.path.join(project_dir, base, self.name))
-                found = os.path.isdir(os.path.join(path, 'conda-meta'))
-                if found or self._path is None:
-                    self._path = path
-                    if found:
-                        break
+            env_paths = os.environ.get('ANACONDA_PROJECT_ENVS_PATH', '').split(os.pathsep)
+
+            # Scan for an existing env with the
+            # same name
+            existing_env = None
+            for base in env_paths:
+                prefix = _prefix(base)
+                if _found(prefix):
+                    existing_env = prefix
+                    break
+
+            # The existing env may be in a read-only directory
+            # which means that we cannot make changes (add/remove packages)
+            if existing_env is not None:
```

Comment:
I think we need to find another place for this logic. I don't think it is wise to execute the _cloning operation itself_ in what is really supposed to be just a rapid property-like call.

Instead, what we need to accomplish is for this clone to occur during the "fix environment deviations" phase. What that also means is that the deviations need to be marked fixable if: 1) a read-write environment can be created in a higher-priority location and 2) cloning is enabled.

Reply:
Agreed. I was thinking along the same lines.

Reply:
I might play with it a bit. It's so close to being ready I'd love to get it into 5.5